DYN-6412 fix performance issue with feature flags, and deadlock with CLI wrapper.

mjkkirschner opened the pull request:
<!--
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.
-->

### Purpose

1. when feature flags does not retrieve data, don't log for every single flag check - this leads to extremely poor performance as WPF tries to update the console window. The console is also guarded by a lock.
2. Modify the CLIWrapper class method `GetData` which reads from standard out of the wrapped processes - so that this method is async, and will timeout after a client controllable timeout.


### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [x] This PR contains no files larger than 50 MB

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. **Mandatory section**

### Reviewers

(FILL ME IN) Reviewer 1 (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

mjkkirschner commented:
@sm6srw I'd like to test this, but we can get it in sooner and followup so civil can consume it.

mjkkirschner commented on src/DynamoUtilities/Md2Html.cs:
```diff
@@ -93,53 +96,63 @@ protected static string GetToolPath(string relativePath)
-        protected virtual string GetData()
+        /// <param name="timeoutms">will return empty string if we don't finish reading all data in the timeout provided in milliseconds.</param>
+        /// <returns></returns>
+        protected virtual async Task<string> GetData(int timeoutms)
```
@sm6srw @pinzart90 I was considering pulling this from a configuration file in the unfortunate case that a user is on a very slow machine, they could have some control over the timeout to avoid never getting in depth documentation or feature flags...

sm6srw replied:
I guess a timeout could happen once in a while on heavy loaded machine. Using a machine where the timeout happens all the time must be a bad experience in general. I think the risk for a timeout is higher for the feature flags as both load and/or network can cause it. Maybe add the value to a file if it is easy? The question then would be one or two values?

pinzart90 replied:
Let's test out the timeouts by blocking the thread on the featureflags exe side.
Maybe we should just have a dedicated config to turn off ff completely and keep the timeout to something sensible by default

mjkkirschner replied:
it's not just FF, but all of these CLIWrapper base tools (like md2html.exe)

`Let's test out the timeouts by blocking the thread on the featureflags exe side.`
I've already tested locally by immediately exiting from the process or somehow stopping std out from being written to - is that what you mean?

I just think it's hard to pick a sensible timeout for all user scenarios though I guess it's not the end of the world if feature flags report their default or the user never gets in depth documentation.

mjkkirschner replied:
discussed with @pinzart90 I've increased these and I think it's more sensible now.

sm6srw commented:
I don't think we need it asap for civil so adding tests is ok.

mjkkirschner commented on src/DynamoUtilities/DynamoFeatureFlagsManager.cs:
```diff
@@ -54,23 +62,27 @@ internal void CacheAllFlags()
 
             //wait for response
-            var dataFromCLI = GetData();
+            var dataFromCLI = GetData(2000).Result;    
```
see question above - I was thinking to define it in a config file to allow users a workaround - is that overkill?

pinzart90 replied:
Sounds like it is too low level

pinzart90 commented on src/DynamoUtilities/Md2Html.cs:
```diff
@@ -93,53 +96,63 @@ protected static string GetToolPath(string relativePath)
+        protected virtual async Task<string> GetData(int timeoutms)
         {
-            if (process.HasExited)
+            var readStdOutTask = Task.Factory.StartNew(() =>
```
Why not [Task.Run](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.run?view=net-8.0) ?

aparajit-pratap commented on src/DynamoUtilities/DynamoFeatureFlagsManager.cs:
```diff
@@ -87,12 +99,21 @@ internal T CheckFeatureFlag<T>(string featureFlagKey, T defaultval)
+                {
+                    case true when !nullCacheLogged:
+                        RaiseMessageLogged("The flags cache is null, something went wrong retrieving feature flags. This message will not be logged again, and future calls to CheckFeatureFlags will return default values.");
+                        nullCacheLogged = true;
```
no resource strings?

mjkkirschner replied:
no, these logs are intended for internal use.

pinzart90 commented on src/DynamoUtilities/DynamoFeatureFlagsManager.cs:
```diff
@@ -87,12 +99,21 @@ internal T CheckFeatureFlag<T>(string featureFlagKey, T defaultval)
-            // and log.
+            // and log once,
            
             if(AllFlagsCache == null)
```
is this thread safe ?

mjkkirschner replied:
no
![Screenshot 2023-11-22 at 1 01 59 PM](https://github.com/DynamoDS/Dynamo/assets/508936/76f6231f-43b7-4bf7-a8cd-12bb0f804121)

pinzart90 commented on src/DynamoUtilities/Md2Html.cs:
```diff
@@ -93,53 +96,63 @@ protected static string GetToolPath(string relativePath)
+                    return writer.ToString();
+                }
+            });
+            var completedTask = await Task.WhenAny(readStdOutTask, Task.Delay(TimeSpan.FromMilliseconds(timeoutms)));
```
so basically we are blocking a threadpool thread for 2 seconds in the worst case scenario ?

mjkkirschner replied:
well yes, we block a thread pool thread, but we are using `Task.Result` so if I'm understanding correctly we're blocking the main thread as well.

mjkkirschner replied:
For FF case, the getData call is actually on a thread pool thread, I think for Md2html it's on the main thread.

mjkkirschner commented:
@pinzart90 @sm6srw @aparajit-pratap please take another look:
1. discussed with Tibi a bit, and I removed even more logging from FF for now.
2. filed a followup to improve our logging system so it can't be overwhelm the UI. Added TODOs inline for this task.
3. added tests and increased the timeouts.

aparajit-pratap commented on src/DynamoUtilities/Md2Html.cs:
It's confusing why the CLIWrapper class is in Md2Html.cs, can we move it out to its own cs file?

mjkkirschner replied:
sure, can do.
